### PR TITLE
Allow higher versions of phpstan to be installed

### LINKED ISSUE #42: Mismatch between requirements
It is not possible to update to magento **2.4.0** community edition because your module requires **phpstan/phpstan:** **^0.12.24**, while magento requires **phpstan/phpstan:** **>=0.12.3 <=0.12.23** (https://packagist.org/packages/magento/community-edition).

### LINKED ISSUE #31176: Update phpstan/phpstan dependency to PHP 8 compatible version
### Description (*)
Update phpstan/phpstan dependency version from ">=0.12.3 <=0.12.23" to ">=0.12.3 <=0.12.54"
Ensure composer install runs on PHP 7.4 and 8.0

### PR DESCRIPTION
**Pull request by hostep**
### Description (*)
See discussions in https://github.com/bitExpert/phpstan-magento/issues/42 and https://github.com/bitExpert/phpstan-magento/pull/43

Basically it boils down to:
- `bitexpert/phpstan-magento` need `phpstan/phpstan` 0.12.26 or higher
- this is possible in Magento 2.3.6: https://github.com/magento/magento2/blob/2.3.6/composer.json#L122
- this is not possible in Magento 2.4.0 or 2.4.1: https://github.com/magento/magento2/blob/2.4.0/composer.json#L95
- problem was introduced in https://github.com/magento/magento2/commit/700de1106e26b03a47e974525bd00bc39f617f8b
- as far as I know, phpstan is developed with semantic versioning in mind, so if they add a breaking change, it will be released as 0.13.0 and then the contraint `^0.12.3` will prevent Magento to use this newer version.

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/462

### Fixed Issues (if relevant)
1. Fixes https://github.com/bitExpert/phpstan-magento/issues/42: Mismatch between requirements
2. Fixes https://github.com/magento/magento2/issues/31176

### Manual testing scenarios (*)
1. Somehow, tell phpstan to run on the Magento 2 source code and it should run without problems (it might detect more/different issues, and that is expected)

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hostep** [comment]
@magento run all tests

(this probably doesn't make much sense, because phpstan is executed on changed source code files and none have been changed in this PR)

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8363](https://jira.corp.magento.com/browse/ENGCOM-8363) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
This was introduced in https://github.com/magento/magento2/commit/700de1106e26b03a47e974525bd00bc39f617f8b
I don't really see any reasons for that, but that need to be checked additionally during a testing of this PR.

**Comment by ihor-sviziev** [comment]
I'm pretty sure failed tests not related to changes in this PR, but let's restart tests
@magento run Functional Tests CE, Functional Tests EE

**Comment by ihor-sviziev** [comment]
**For QA** - unfortunately all the time someone changes composer.json - we have a conflict in composer.lock in "hash".
Unfortunately processing PR is quite slow process, so it's almost impossible to test just merged version of composer.lock.
Could you fix the conflict yourself by accepting any of versions of "hash" and then running `composer update --lock`?

**Comment by hws47a** [comment]
I had a case when PHPStan was failing with Magento core configuration with PHPStan version >=0.12.29 and definitely on 0.12.32. So please be careful with the update as it can break tests. I'm sure there was a reason why the higher band was introduced.

**Comment by ihor-sviziev** [comment]
@hws47a if you could give us steps to reproduce - it will be really helpful

**Comment by hostep** [comment]
@fascinosum: do you happen to remember why the upper constraint for version 0.12.23 was added in https://github.com/magento/magento2/commit/700de1106e26b03a47e974525bd00bc39f617f8b? Thanks!

**Comment by hostep** [comment]
Rebased on top of 2.4-develop, solved conflicts, further updated `phpstan/phpstan` to 0.12.63 and force pushed my commit to keep a clean history.

@magento run all tests

**Comment by sivaschenko** [comment]
Thanks for the pull request @hostep ! Sorry for the delay in processing, I've added this PR to PHP 8 project. The pull request will most likely will be processed and merged early in January 2021

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

@sivaschenko thank you so much for describing your plan and give us some reasonable expecations

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by hostep** [comment]
@sivaschenko: I read in https://github.com/magento/magento2/issues/31081 that PHP 8 support is coming to Magento 2.5, but this PR was never created with the intention of solving a PHP 8 compatibility issue. I'd rather see this being merged in `2.4-develop` so this can get released in Magento 2.4.3.
Can you confirm if that can happen?

**Comment by sivaschenko** [comment]
Hi @hostep we have a dedicated branch for PHP 8 compatibility and dependency updates (that are BIC so cannot be merged to 2.4 without approval). Depending on the amount of backward incompatible changes in php8-develop branch we are going to decide if it should be merged to 2.4 or 2.5. More information here: https://community.magento.com/t5/Magento-DevBlog/Magento-PHP-8-Compatibility-Community-Project/ba-p/464983

**Comment by hostep** [comment]
Thanks!

So why is this PR considered BiC ?

In my opinion limiting the highest version of phpstan introduced in 2.4.0 itself was a BIC change with Magento 2.3.6 where there was no upper limit 😉

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by fascinosum** [comment]
> @fascinosum: do you happen to remember why the upper constraint for version 0.12.23 was added in [700de11](https://github.com/magento/magento2/commit/700de1106e26b03a47e974525bd00bc39f617f8b)? Thanks!

hi @hostep,
sorry for the delayed response. The latest versions had BIC changes and caused our static tests to fail due to dependency on a low-level implementation of this library

**Comment by ihor-sviziev** [comment]
@fascinosum so far, static tests now passing, so the issue isn't reproducing anymore with the latest version? Or are there some specific steps to reproduce?

**Comment by hostep** [comment]
@ihor-sviziev: I believe the static tests are only executed against changed files in a PR, so in this case against the two composer files, but phpstan doesn't have much to do with that, so I guess we need to test this in some other way, but I don't know how.

@fascinosum: thanks for the reply, but do you have more concrete information? How can we test if that issue you ran against no longer occurs with a newer version of phpstan?

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by hostep** [comment]
@ihor-sviziev: if I need to rebase this PR on the latest `php8-develop` branch in order to try to hopefully solve these flaky automated tests, let me know.

**Comment by ihor-sviziev** [comment]
@hostep yes, please

**Comment by hostep** [comment]
Ok done:
- rebased on top of `php8-develop`
- further upgraded phpstan from 0.12.63 to 0.12.69

However, it looks like the `php8-develop` branch is pretty old, it was last updated 26 days ago (at the moment of writing), shouldn't this branch be kept more up2date with `2.4-develop`? @sivaschenko

@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@hostep php8-develop branches update is in progress now

**Comment by hostep** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by sivaschenko** [comment]
Added related ee pull request

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sivaschenko, could you also resolve the conflict by accepting any version of "hash" and executing `composer update --lock` (it will generate a new hash)?

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by fascinosum** [comment]
hi @hostep , @ihor-sviziev 
see steps to reproduce below (tested on 2.3 and 2.4)

- Run the command in console 
`echo 'setup/src/Magento/Setup/Test/Unit/Console/Style/MagentoStyleTest.php' >> $magento_DIR$/dev/tests/static/testsuite/Magento/Test/_files/changed_files_local.txt`
**_Replace $magento_DIR$ with the real path to the Magento installation directory_**

- run
`composer require --dev phpstan/phpstan:0.12.24` or just use some other later version
 
- run \Magento\Test\Php\LiveCodeTest::testPhpStan

cc: @sivaschenko
Internal ticket was created MC-40914

**Comment by hostep** [comment]
Thanks @fascinosum !

Can you tell us how we need to execute `\Magento\Test\Php\LiveCodeTest::testPhpStan` ?
Is there an easy way to do this?

**Comment by fascinosum** [comment]
@hostep, just run it on your local
 https://github.com/magento-commerce/magento2ce/blob/10a70656165765af1f9a2277eb91410c79eacd33/dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php#L449 in PHPStorm
 using https://github.com/magento-commerce/magento2ce/blob/2.4-develop/dev/tests/static/phpunit.xml.dist configuration for PHPUnit. This is the easiest way how to run and debug
 If you are not using PHPStorm, you can run a CLI command for PHPUnit

**Comment by hostep** [comment]
Ok great!

So something like this on the CLI I believe:
```
vendor/bin/phpunit -c dev/tests/static/phpunit.xml.dist --testsuite 'PHP Coding Standard Verification' --filter testPhpStan
```

However, I'm still unable to trigger any kind of error:

```
$ composer require --dev phpstan/phpstan:0.12.24
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating phpstan/phpstan (0.12.23 => 0.12.24): Downloading (100%)
...

$ echo 'setup/src/Magento/Setup/Test/Unit/Console/Style/MagentoStyleTest.php' > ./dev/tests/static/testsuite/Magento/Test/_files/changed_files_local.txt

$ vendor/bin/phpunit -c dev/tests/static/phpunit.xml.dist --testsuite 'PHP Coding Standard Verification' --filter testPhpStan
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.

.                                                                                                                                                                                                                                 1 / 1 (100%)

Time: 00:00.215, Memory: 4.00 MB

OK (1 test, 1 assertion)
```

Am I still doing something wrong @fascinosum?

**Comment by hostep** [comment]
Ok, I'm able to trigger it, finally.
I had to hardcode the method `Magento\Test\Php\LiveCodeTest::getWhitelist` to return:
```
return ['setup/src/Magento/Setup/Test/Unit/Console/Style/MagentoStyleTest.php'];
```

After that I can trigger the test, but my god, this is such an awful way to run static tests, I really hope something gets done to this in the future so we have a much more user-friendly way of triggering these, and then also finally [document](https://github.com/magento/devdocs/issues/4495) how you should do it.

----

Anyway, the results are in:

phpstan 0.12.23:
```
$ vendor/bin/phpunit -c dev/tests/static/phpunit.xml.dist --testsuite 'PHP Coding Standard Verification' --filter testPhpStan
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.

.                                                                                                                                                                                                                                 1 / 1 (100%)

Time: 00:02.147, Memory: 4.00 MB

OK (1 test, 1 assertion)
```

phpstan 0.12.24:
```
$ vendor/bin/phpunit -c dev/tests/static/phpunit.xml.dist --testsuite 'PHP Coding Standard Verification' --filter testPhpStan
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.


In Resolver.php line 402:

  Service 'errorFormatter.filtered' (type of Magento\PhpStan\Formatters\FilteredErrorFormatter): Unable to pass specified arguments to PHPStan\Command\ErrorFormatter\TableErrorFormatter::__construct().


analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--generate-baseline [GENERATE-BASELINE]] [--memory-limit MEMORY-LIMIT] [--xdebug] [--] [<paths>...]

F                                                                                                                                                                                                                                 1 / 1 (100%)

Time: 00:00.902, Memory: 6.00 MB

There was 1 failure:

1) Magento\Test\Php\LiveCodeTest::testPhpStan
PHPStan command run failed.
Failed asserting that 1 matches expected 0.

dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php:478

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```

phpstan 0.12.69 (which was used and approved in this PR):
```
$ vendor/bin/phpunit -c dev/tests/static/phpunit.xml.dist --testsuite 'PHP Coding Standard Verification' --filter testPhpStan
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.


In Resolver.php line 442:

  Service 'errorFormatter.github' (type of PHPStan\Command\ErrorFormatter\GithubErrorFormatter): Multiple services of type PHPStan\Command\ErrorFormatter\TableErrorFormatter found: errorFormatter.filtered, errorFormatter.table (needed
  by $tableErrorformatter in __construct())


In Autowiring.php line 50:

  Multiple services of type PHPStan\Command\ErrorFormatter\TableErrorFormatter found: errorFormatter.filtered, errorFormatter.table


analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--generate-baseline [GENERATE-BASELINE]] [--memory-limit MEMORY-LIMIT] [--xdebug] [--fix] [--watch] [--pro] [--] [<paths>...]

F                                                                                                                                                                                                                                 1 / 1 (100%)

Time: 00:01.177, Memory: 6.00 MB

There was 1 failure:

1) Magento\Test\Php\LiveCodeTest::testPhpStan
PHPStan command run failed.
Failed asserting that 1 matches expected 0.

dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php:478

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```

So indeed, something will need to be done inside Magento's code of how it calls phpstan to make it compatible again somehow. I'm not going to put time in this in the next few days I think and we should probably now also put https://github.com/magento/magento2/pull/31972 on hold until a solution is found for this.

**Comment by hostep** [comment]
The easiest way to trigger the newer phpstan version incompatibility is by simply executing:

```
$ ./vendor/bin/phpstan analyse --level 1 --no-progress --memory-limit=4G --configuration ./dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/phpstan.neon setup/src/Magento/Setup/Test/Unit/Console/Style/MagentoStyleTest.php

In Resolver.php line 442:

  Service 'errorFormatter.github' (type of PHPStan\Command\ErrorFormatter\GithubErrorFormatter): Multiple services of type PHPStan\Command\ErrorFormatter\TableErrorFormatter found: errorFormatter.filtered, errorFormatter.table (needed
  by $tableErrorformatter in __construct())


In Autowiring.php line 50:

  Multiple services of type PHPStan\Command\ErrorFormatter\TableErrorFormatter found: errorFormatter.filtered, errorFormatter.table


analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--generate-baseline [GENERATE-BASELINE]] [--memory-limit MEMORY-LIMIT] [--xdebug] [--fix] [--watch] [--pro] [--] [<paths>...]
```

That way you don't need to go through this phpunit mess.

**Comment by fascinosum** [comment]
thank you @hostep,
the CLI command in PHPStorm is
`/usr/local/bin/php $magento_DIR$/vendor/phpunit/phpunit/phpunit --configuration $magento_DIR$/dev/tests/static/phpunit.xml.dist --filter "/::testPhpStan( .*)?$/" --test-suffix LiveCodeTest.php $magento_DIR$/dev/tests/static/testsuite/Magento/Test/Php --teamcity`
All paths are absolute
JFYI: we only run static tests on change files, so you need this hack with `changed_files_local.txt`. But it should work without mocked methods

**Comment by ihor-sviziev** [comment]
It would be good to create a new issue fr fixing this new behavior after the phpstan update as it was recently merged to the php8 branch

**Comment by hostep** [comment]
I'm assuming a core Magento dev wil pick this up since internal ticket MC-40914 was created. It has something to do with that [custom FilteredErrorFormatter class](https://github.com/magento/magento2/blob/800acfaecc9dc18007095c0c8d68934c92337a5e/dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/phpstan.neon#L46-L52) which is custom written for Magento

**Comment by sivaschenko** [comment]
@hostep @ihor-sviziev @fascinosum can you please take a look at this proposed fix: https://github.com/magento/magento2/pull/32215

**Comment by ihor-sviziev** [comment]
@sivaschenko, TBH, I didn't work enough with phpstan to check if your changes are good enough, but from the first look - looks good.

**Comment by hws47a** [comment]
Good job @sivaschenko

**Pull request merged**
